platform/kube/inject/initializer.go

```diff
@@ -0,0 +1,389 @@
+			SidecarProxyUID:   DefaultSidecarProxyUID,
+			EnableCoreDump:    true,
+			Version:           version.Line(),
+			Mesh:              mesh,
+			MeshConfigMapName: "istio",
+		},
+	}
+
+	kinds := []struct {
+		resource   string
+		getter     cache.Getter
+		objType    runtime.Object
+		initialize func(in, out interface{}) error
+	}{
+		{
+			"deployments",
+			i.clientset.AppsV1beta1().RESTClient(),
+			&appsv1beta1.Deployment{},
+			i.initializeDeployment,
+		},
+		{
+			"statefulsets",
+			i.clientset.AppsV1beta1().RESTClient(),
+			&appsv1beta1.StatefulSet{},
+			i.initializeStatefulSet,
+		},
+		{
+			"jobs",
+			i.clientset.BatchV1().RESTClient(),
+			&batchv1.Job{},
+			i.initializeJob,
+		},
+		{
+			"daemonsets",
+			i.clientset.ExtensionsV1beta1().RESTClient(),
+			&v1beta1.DaemonSet{},
+			i.initializeDaemonSet,
+		},
+		{
+			"replicasets",
+			i.clientset.ExtensionsV1beta1().RESTClient(),
+			&v1beta1.ReplicaSet{},
+			i.initializeReplicaSet,
+		},
+		{
+			"replicationcontrollers",
+			i.clientset.CoreV1().RESTClient(),
+			&v1.ReplicationController{},
+			i.initializeReplicationController,
+		},
+		{
+			"pods",
+			i.clientset.CoreV1().RESTClient(),
+			&v1.Pod{},
+			i.initializePod,
+		},
+	}
+
+	for _, kind := range kinds {
+		kind := kind // capture the current value of `kind`
+
+		watchList := cache.NewListWatchFromClient(kind.getter, kind.resource,
+			i.options.Namespace, fields.Everything())
+
+		// Wrap the returned watchlist to workaround the inability to include
+		// the `IncludeUninitialized` list option when setting up watch clients.
+		includeUninitializedWatchList := &cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.IncludeUninitialized = true
+				return watchList.List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.IncludeUninitialized = true
+				return watchList.Watch(options)
+			},
+		}
+
+		_, controller := cache.NewInformer(includeUninitializedWatchList, kind.objType, i.options.ResyncPeriod,
+			cache.ResourceEventHandlerFuncs{
+				AddFunc: func(in interface{}) {
+					obj, err := meta.Accessor(in)
+					if err != nil {
+						return
+					}
+					if !i.hasIstioInitializerNext(obj) {
+						return
+					}
+					out, err := runtime.NewScheme().DeepCopy(in)
+					if err != nil {
+						return
+					}
+					glog.Infof("Initializing %v: %v", meta.AsPartialObjectMetadata(obj).Kind, obj.GetName())
+
+					if err := kind.initialize(in, out); err != nil {
+						glog.Errorf("Could not initialize %s: %v", kind.resource, err)
+					}
+				},
+			},
+		)
+		i.controllers = append(i.controllers, controller)
+	}
+	return i
+}
+
+func (i *Initializer) hasIstioInitializerNext(object metav1.Object) bool {
+	glog.V(2).Infof("ObjectMeta initializer info %v/%v policy:%q status:%q %v",
+		object.GetNamespace(), object.GetName(),
+		object.GetAnnotations()[istioSidecarAnnotationPolicyKey],
```

Comment:
Good idea @costinm. I created https://github.com/istio/pilot/issues/1061 to track that FR.